File: workloads/_template/tools_example.cpp
/**
 * Example C++ MCP Tool definition for Arm workload auto-discovery.
 * Compile step generates build/mcp_schemas/cpp_tool_example.json sidecar.
 */

#include <iostream>
#include <vector>

// Macro annotation for C++ MCP Tool registration
#define ARM_MCP_TOOL(name, description)

// @mcp_tool: Optimizes 2D Depthwise Separable Convolution using Arm KleidiAI intrinsics
ARM_MCP_TOOL(kleidiai_depthwise_conv2d, "Optimizes 2D Depthwise Separable Convolution using Arm KleidiAI micro-kernels.")
void kleidiai_depthwise_conv2d(const float* input, float* output, int width, int height, int channels) {
    // Arm KleidiAI optimized kernel implementation
    std::cout << "Executing Arm KleidiAI depthwise conv2d for " << width << "x" << height << " tensor." << std::endl;
}

int main() {
    std::cout << "C++ Arm Kernel Template Initialized." << std::endl;
    return 0;
}

File: workloads/_template/ToolsExample.java
package com.arm.developer.workloads;

/**
 * Example Java MCP Tool definition for Arm workload auto-discovery.
 * Annotation processor emits build/mcp_schemas/java_tool_example.json sidecar.
 */
public class ToolsExample {

    public @interface McpTool {
        String name() default "";
        String description() default "";
    }

    @McpTool(
        name = "java_executorch_tensor_norm",
        description = "Performs FP32 Tensor Normalization for ExecuTorch mobile pipelines on Arm Cortex-X cores."
    )
    public static float[] javaExecuTorchTensorNorm(float[] inputData, float mean, float stdDev) {
        float[] normalized = new float[inputData.length];
        for (int i = 0; i < inputData.length; i++) {
            normalized[i] = (inputData[i] - mean) / stdDev;
        }
        return normalized;
    }

    public static void main(String[] args) {
        System.out.println("Java Arm Workload Template Initialized.");
    }
}
